transport/src/main/java/io/netty/channel/SingleThreadEventLoop.java

```diff
@@ -71,11 +86,82 @@ public ChannelFuture register(final Channel channel, final ChannelPromise promis
         return promise;
     }
 
+    /**
+     * Adds a task to be run once at the end of next (or current) {@code eventloop} iteration.
+     *
+     * @param task to be added.
+     */
+    final void onEventLoopIteration(Runnable task) {
+        ObjectUtil.checkNotNull(task, "task");
+        if (isShutdown() || isShuttingDown() || isTerminated()) {
+            reject();
+        }
+
+        if (!tailTasks.offer(task)) {
+            throw new RejectedExecutionException("Task queue for onEventLoopIteration tasks is full.");
+        }
+
+        if (wakesUpForTask(task)) {
+            wakeup(inEventLoop());
+        }
+    }
+
+    /**
+     * Removes a task that was added previously via {@link #onEventLoopIteration(Runnable)}.
+     *
+     * @param task to be removed.
+     *
+     * @return {@code true} if the task was removed as a result of this call.
+     */
+    final boolean removeOnEventLoopIterationTask(Runnable task) {
+        return tailTasks.remove(ObjectUtil.checkNotNull(task, "task"));
+    }
+
     @Override
     protected boolean wakesUpForTask(Runnable task) {
         return !(task instanceof NonWakeupRunnable);
     }
 
+    @Override
+    protected boolean runAllTasks(long timeoutNanos) {
+        assert inEventLoop();
+        runTimeout = timeoutNanos;
+        if (lastRunTimedOut) {
```

Comment:
I wonder if we can avoid the extra complexity of `tailTasks` and related timing issues? Seems like we only expect to have a single Runnable at a time, and this Runnable is only appended as a result of a Runnable being run from `taskQueue`.  Could we instead have a member variable (e.g. `Runnable autoFlushRunner` (may need to be volatile/atomic)) in `SingleThreadEventExecutor` and run it after `runAllTasks` completes (assuming its not null)? It may be easier this way to run it even if a "timeout" of regular tasks occurs potentially to ensure a flush is done before the eventloop goes back to the selector (if we wanted to).

In addition to the timeout complexities there may be other API ambiguities we have to deal with. For example with the current approach `SingleThreadEventLoop` overrides `hasTasks()` but not `pendingTasks()` (and some other queue related methods). Which we should also be able to avoid these types of API ambiguities if we avoid adding an additional layer of queuing and narrow our focus.